Fj qm allocation

MizuBishi opened the pull request:
Hoi Joel 

Anbei die Wahl der QM für den Fachcoach, ist aber noch nicht ganz fertig. 

LG Michelle

MizuBishi committed:
Edited existing fields.

MizuBishi committed:
Added autoselect for QM.

MizuBishi committed:
Added Autosuggest for QM and shows Name and Email of QM.

MizuBishi committed:
Solved problem with initial State. Now you get an empty field when you "add" a new project and not the selected QM, from the project before.

MizuBishi committed:
Edited entries in Status and css adaptions.

MizuBishi commented:
Hoi Joel, bitte noch nicht gegenprüfen. Ich muss zuerst noch ein paar Änderungen vornehmen und gebe dir dann Bescheid. 
LG Michelle

MizuBishi committed:
Fix compile error backend.

MizuBishi committed:
Renamed EditProject -> CreateProject.

MizuBishi committed:
Simplified CreateProjectPage & Project entity.

Project entity changes:
- Status is either OPEN or CLOSE based on stop=null

MizuBishi committed:
Updated ProjectOverview

MizuBishi committed:
CreateProjectPage : QM : AutoSuggest <-> text; added .members.student.email handling to POST /projects

MizuBishi committed:
Added ShowProjectPage

MizuBishi commented:
Hoi Joel 
Pull Request ist fertig. Ich habe es momentan wie folgt implementiert:
- Projekte adden 
- anschliessend können die Projekte nicht mehr verändert werden bis auf das End date (Folgeprojekt?)
- Sobald das End Date eingegeben und erreicht wurde, ist der Projektstatus auf closed.
- Projekte können nachträglich nicht mehr geöffnet werden (Folgeprojekt?)

joelmeiller commented on src/main/java/ch/fhnw/p2p/controller/ProjectController.java:
```diff
@@ -83,10 +90,7 @@
 		}
-		oldProject.setTitle(project.getTitle());
-		oldProject.setStart(project.getStart());
 		oldProject.setStop(project.getStop());
```
ist nur erlaubt wenn das Projekt nicht im Status FINAL ist. Könnte man mit einem if noch abfangen

MizuBishi replied:
neu : projekt kann nur geschlossen werden wenn status vorher auf FINAL war (siehe doku 5.3.3)

joelmeiller commented on src/main/java/ch/fhnw/p2p/controller/ProjectController.java:
```diff
@@ -107,11 +112,20 @@
 			return new ResponseEntity<HttpStatus>(HttpStatus.PRECONDITION_FAILED);
 		}
+		if (project.getMembers().size() > 0) {
+			User qmUser = userRepo.findByEmail(project.getMembers().iterator().next().getStudent().getEmail()).get();
```
Hier müsste theoretisch noch abgefragt werden, ob der User nicht bereits in einem Projekt ist. Also

`userRepo.findByEmailAndStatus(...., User.Status.FREE)`

MizuBishi replied:
erledigt

joelmeiller commented on src/main/java/ch/fhnw/p2p/controller/ProjectController.java:
```diff
@@ -107,11 +112,20 @@
 			logger.debug("Successfully saved project[" + project.getId() + "]: " + project.toString());
 			return new ResponseEntity<HttpStatus>(HttpStatus.OK);
 		} catch (Exception e) {
+			e.printStackTrace();
```
habe ich mit `logger.error(e)` bei mir implementiert, so dass es auch im log-File erscheint

joelmeiller commented on src/main/java/ch/fhnw/p2p/controller/ProjectController.java:
```diff
@@ -107,11 +112,20 @@
+			project.getMembers().clear();
+			Role qmRole = roleRepo.findByShortcut("QM");
+			Member qmMember = new Member(project, qmUser, qmRole);
+			project.getMembers().add(qmMember);
```
besser wäre hier die Methode updateProjectMembers aufzurufen, da diese das alles schon implementiert. Können wir aber noch anschauen.

MizuBishi replied:
methode nicht gefunden...

joelmeiller commented on src/main/java/ch/fhnw/p2p/entities/Project.java:
```diff
@@ -31,23 +34,40 @@
+	
+	public static enum Status {
+		OPEN,
+		CLOSE,
```
Ich muss mir das noch überlegen. Ich brauche den Status FINAL für den Ratingablauf, aber vielleicht können wir tatsächlich den Status auf diese 2 zurückkürzen. Muss ich noch anschauen

MizuBishi replied:
FINAL wieder hinzugefügt.

joelmeiller commented on src/main/java/ch/fhnw/p2p/entities/Project.java:
```diff
@@ -72,7 +90,7 @@ public Project() {
 		this();
 		this.title = title;
-		this.slug = Slug.makeSlug(title);
+//		this.slug = Slug.makeSlug(title);
```
warum entfernt?

MizuBishi replied:
funktioniert im moment nicht richtig (nicht eindeutig); auf später verschoben...

joelmeiller commented on src/middleware/project.js:
```diff
@@ -11,18 +11,21 @@ export const getProject = (id, callback) =>
       ...data,
       start: new Date(data.start),
-      stop: new Date(data.stop),
+      stop: data.stop == null ? null : new Date(data.stop),
```
das sollte nicht funktionieren oder? Es müssen 3 === Zeichen sein???

MizuBishi replied:
geht auch mit zwei (undefined == null != new Date())

abgeändert.

joelmeiller commented on src/middleware/project.js:
```diff
@@ -11,18 +11,21 @@ export const getProject = (id, callback) =>
   start: project.start.toISOString().substring(0, 10),
-  stop: project.stop.toISOString().substring(0, 10),
-  zeitmodell: project.zeitmodell,
+  stop: project.stop == null ? null : project.stop.toISOString().substring(0, 10),
```
dito

joelmeiller commented on src/ui/pages/ProjectPage.jsx:
```diff
@@ -12,14 +12,16 @@ import {
   TableRowColumn,
 } from 'material-ui/Table';
 
+const dateToTerm = date => (date.getMonth() + 1 >= 9 || date.getMonth() + 1 <= 2 ? 'HS' : 'FS') + date.getYear();
```
state of the art ist hier moment.js

MizuBishi replied:
moment versteht leider keine semester; habe middleware/project.js + pages/ShowProjectPage.jsx abgeändert.

joelmeiller commented:
Hoi Michelle,

Da hast du ja ordentlich was gemacht :) Sieht alles in allem super aus. Ich habe noch ein paar Kommentare, aber die kannst du ignorieren. Das einzige was mich verwirrt sind die 2 anstatt 3 === Zeichen bei der Abfrage `stop == null` in den javaScript files. Kannst du das noch kurz anschauen, dann kannst du den Pull-Request mergen

MizuBishi reviewed (commented):
merci für die tips. das meiste implmentiert wie vorgeschlagen, slug auf später verschoben.

MizuBishi committed:
Veränderungen wie auf GitHub im pull request vorgeschlagen.

Status ist jetzt OPEN (noch nicht bereit) FINAL (kann abgeschlossen werden) CLOSE (abgeschlossen).

MizuBishi committed:
Fehlerkontrolle Kommunikation backend (getProjectList, getTeam, getUserSettings)

MizuBishi committed:
Backend überprüft jetzt neues Projekt (mappings.NewProject) + UI toasts für verschiedene Fehler.

- BaseController übernimmt das ExceptionHandling
- NewProject mit neuem Feld emailQm
- Error checks : Verbindungsproblem, JSON error, Fehler mit ApiResponse.message
- PUT /api/projects/{id} -> /api/projects/{id}/status

MizuBishi committed:
Merge branch 'master' into fj-qm-allocation

Manuelles Testing:
- Fachjury-Bereich
- QM (oberflächlich)
- TM (oberflächlich)

* master: (24 commits)
  adjusted fixtures for publish on server
  updated status handling for TM to accept deviation / added functionality for QM to close project
  correct member rating overview / adjust routing
  corrections for TeamRatingPage after project is final / eslint improvements
  Status handling adjustments / First draft of deviation correction for QM
  Status correction
  Adjustments for add/remove criteria --> update team member ratings accordingly
  Adjusted remove member method to remove ratings from other members
  adjusted EqualsAndHashCode to include id in entities / corrected saveTeam functionality / adjusted layout
  added template rating to team member ratings
  implemented deviation calculator
  Corrections in progress & rating calculation and display logic
  Member.clone() correction
  Correction for duplicates in MemberRating (List to Set)
  Adjusted MyRating views to MemberRating views and started with proper data mapping
  Rating adjustments for QM to complete its own ratings for the other Team Members
  TM functionality is ready for User Testing
  Corrected inbox handling and adjusted UX to force user to approve its project allocation before it can change ratings
  status handling corrections
  status corrections for user settings and dependency handling
  ...

Conflicts:
	src/actions/projectList.js
	src/main/java/ch/fhnw/p2p/controller/MemberRatingController.java
	src/main/java/ch/fhnw/p2p/controller/ProjectController.java
	src/main/java/ch/fhnw/p2p/entities/Project.java
	src/ui/root/routes.jsx

MizuBishi commented:
Habe einen neuen Pull-Request gemacht im gleichen Branch.
